parl/remote/decorator.py

```diff
@@ -0,0 +1,203 @@
+            message = socket.recv_string()
+            logger.info("[connect_server] done, message from server:{}".format(
+                message))
+            self.connect_socket = socket
+
+        def __getattr__(self, attr):
+            """
+            Call the function of the initial class. The wrapped class do not have 
+            same functions as the unwrapped one. 
+            We have to call the function of the function in unwrapped class,
+            This implementation utilise a function wrapper.
+            """
+            if hasattr(self.unwrapped, attr):
+
+                def wrapper(*args, **kw):
+                    return getattr(self.unwrapped, attr)(*args, **kw)
+
+                return wrapper
+            raise AttributeError(attr)
+
+        def remote_run(self, server_ip, server_port):
+            """
+            connect server and wait for requires of running functions from server side.
+
+            Args:
+                server_ip(str): server's ip
+                server_port(int): server's port
+            """
+            self.connect_server(server_ip, server_port)
+            while True:
+                function_name = self.reply_socket.recv_string()
+                self.reply_socket.send_string("OK")
+                data = self.reply_socket.recv()
+                args, kw = loads_argument(data)
+                ret = getattr(self.unwrapped, function_name)(*args, **kw)
+                ret = dumps_return(ret)
+                self.reply_socket.send(ret)
+
+    class ServerWrapper(object):
+        """
+        Wrapper for remote class at server side. 
+        """
+
+        def __init__(self, *args):
+            """
+            Args:
+                args: arguments used to initialize the initial class
+            """
+            self.unwrapped = (cls(*args)).unwrapped
+            self.command_socket = None
+            self.internal_lock = threading.Lock()
+
+        def connect_client(self, client_info):
+            """
+            build another connection with the client to send command to the client.
+            """
+            client_address, client_id = client_info.split(' ')
+            context = zmq.Context()
+            socket = context.socket(zmq.REQ)
+            logger.info(
+                "[connect_client] client_address:{}".format(client_address))
+            socket.connect("tcp://{}".format(client_address))
+            self.command_socket = socket
+            self.client_id = client_id
+
+        def __getattr__(self, attr):
+            """
+            Run the function at client side. we also implement this through a wrapper.
+
+            Args:
+                attr(str): a function name specify which function to run.
+            """
+            if hasattr(self.unwrapped, attr):
+
+                def wrapper(*args, **kw):
+                    self.internal_lock.acquire()
+                    self.command_socket.send_string(attr)
```

Comment:
Should we use one send here? Like use send_multipart([function_name, args]), instead of sending twice.

Reply:
Yes, it would be faster if send_multipart was used. The reason I write in this way is to make it easier for us to transfer our connection mode to publish&require mod,  as connecting twice is required. in publish&require mode.